src/main/java/frc/robot/vision/sources/LimeLightSource.java

```diff
@@ -0,0 +1,153 @@
+package frc.robot.vision.sources;
+
+import edu.wpi.first.math.Pair;
+import edu.wpi.first.math.geometry.Pose3d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Rotation3d;
+import edu.wpi.first.networktables.NetworkTableEntry;
+import edu.wpi.first.networktables.NetworkTableInstance;
+import frc.robot.subsystems.GBSubsystem;
+import frc.robot.vision.GyroAngleValues;
+import frc.robot.vision.filters.Filter;
+import frc.robot.vision.rawdata.RawVisionAprilTagData;
+import frc.robot.vision.VisionConstants;
+import frc.robot.vision.limelights.LimelightEntryValue;
+import frc.utils.Conversions;
+import frc.utils.alerts.Alert;
+import frc.utils.alerts.AlertManager;
+import frc.utils.alerts.PeriodicAlert;
+import frc.utils.time.TimeUtils;
+
+import java.util.Optional;
+
+public class LimeLightSource extends GBSubsystem implements RobotPoseEstimatingVisionSource {
+
+	private final NetworkTableEntry robotPoseEntry;
+	private final NetworkTableEntry oldRobotPoseEntry;
+	private final NetworkTableEntry aprilTagIdEntry;
+	private final NetworkTableEntry aprilTagPoseEntry;
+	private final NetworkTableEntry robotOrientationEntry;
+	private final String name;
+	private Filter<RawVisionAprilTagData> filter;
+	private double[] robotPoseArray;
+	private double[] aprilTagPoseArray;
+	private Rotation2d robotHeading;
+	private GyroAngleValues gyroAngleValues;
+	private boolean useOldRobotPoseEntry;
+
+	public LimeLightSource(String name, String parentLogPath, Filter<RawVisionAprilTagData> filter) {
+		super(parentLogPath + name + "/");
+
+		this.name = name;
+		this.filter = filter;
+		this.robotPoseEntry = getLimelightNetworkTableEntry("botpose_orb_wpiblue");
+		this.oldRobotPoseEntry = getLimelightNetworkTableEntry("botpose_wpiblue");
+		this.aprilTagPoseEntry = getLimelightNetworkTableEntry("targetpose_cameraspace");
+		this.aprilTagIdEntry = getLimelightNetworkTableEntry("tid");
+		this.robotOrientationEntry = getLimelightNetworkTableEntry("robot_orientation_set");
+		this.gyroAngleValues = new GyroAngleValues(Rotation2d.fromDegrees(0), 0, Rotation2d.fromDegrees(0), 0, Rotation2d.fromDegrees(0), 0);
+		this.useOldRobotPoseEntry = false;
+
+		AlertManager.addAlert(
+			new PeriodicAlert(
+				Alert.AlertType.WARNING,
+				this.getLogPath() + "DisconnectedAt",
+				() -> getLimelightNetworkTableEntry("tv").getInteger(-1) == -1
+			)
+		);
+	}
+
+	public void switchToOldBotPose(boolean useOldRobotPose) {
+		useOldRobotPoseEntry = useOldRobotPose;
+	}
+
+	@Override
+	public void updateGyroAngles(GyroAngleValues gyroAngleValues) {
+		this.gyroAngleValues = gyroAngleValues;
+	}
+
+	@Override
+	public void updateEstimation() {
+		robotOrientationEntry.setDoubleArray(
+			new double[] {
+				gyroAngleValues.yaw().getDegrees(),
+				gyroAngleValues.yawRate(),
+				gyroAngleValues.pitch().getDegrees(),
+				gyroAngleValues.pitchRate(),
+				gyroAngleValues.roll().getDegrees(),
+				gyroAngleValues.rollRate()}
+		);
```

Comment:
How sure are we about the values of the angles' rates?

Reply:
About what in them? The unit is degrees/second. I updated the GyroAngleValues record's javadoc the reflect that.